Error reading Landsat Data off S3

This is either a Bug or a confusing error message:

<img width="1131" alt="screen shot 2017-08-12 at 7 37 01 pm" src="https://user-images.githubusercontent.com/2320142/29244935-dac7b19e-7f95-11e7-9c5e-a49720286606.png">

Code:
```python
import geopyspark as gps

from pyspark import SparkContext
# Set up our spark context
conf = gps.geopyspark_conf(appName="Landsat") \
          .setMaster("local[*]") \
          .set(key='spark.ui.enabled', value='true') \
          .set(key="spark.driver.memory", value="8G")
sc = SparkContext(conf=conf)
target_image = "s3://landsat-pds/c1/L8/138/217/LC08_L1GT_138217_20170703_20170703_01_RT/LC08_L1GT_138217_20170703_20170703_01_RT_B5.TIF"
rdd = gps.geotiff.get(gps.LayerType.SPATIAL, target_image, max_tile_size=512, num_partitions=50)
tiled_layer = rdd.tile_to_layout(gps.GlobalLayout(), target_crs="EPSG:3857")
```

Full stack trace:
```
---------------------------------------------------------------------------
Py4JJavaError                             Traceback (most recent call last)
<ipython-input-12-9d091f0366da> in <module>()
----> 1 tiled_layer = rdd.tile_to_layout(gps.GlobalLayout(), target_crs="EPSG:3857")

/home/hadoop/.local/lib/python3.4/site-packages/geopyspark/geotrellis/layer.py in tile_to_layout(self, layout, target_crs, resample_method)
    644         if target_crs:
    645             target_crs = crs_to_proj4(target_crs)
--> 646             return _reproject(target_crs, layout, resample_method, self)
    647 
    648         if isinstance(layout, Metadata):

/home/hadoop/.local/lib/python3.4/site-packages/geopyspark/geotrellis/layer.py in _reproject(target_crs, layout, resample_method, layer)
    108 
    109     if isinstance(layout, (LocalLayout, GlobalLayout, LayoutDefinition)):
--> 110         srdd = layer.srdd.reproject(target_crs, layout, resample_method)
    111         return TiledRasterLayer(layer.layer_type, srdd)
    112 

/usr/local/spark-2.1.0-bin-hadoop2.7/python/lib/py4j-0.10.4-src.zip/py4j/java_gateway.py in __call__(self, *args)
   1131         answer = self.gateway_client.send_command(command)
   1132         return_value = get_return_value(
-> 1133             answer, self.gateway_client, self.target_id, self.name)
   1134 
   1135         for temp_arg in temp_args:

/usr/local/spark-2.1.0-bin-hadoop2.7/python/lib/py4j-0.10.4-src.zip/py4j/protocol.py in get_return_value(answer, gateway_client, target_id, name)
    317                 raise Py4JJavaError(
    318                     "An error occurred while calling {0}{1}{2}.\n".
--> 319                     format(target_id, ".", name), value)
    320             else:
    321                 raise Py4JError(

Py4JJavaError: An error occurred while calling o37.reproject.
: org.apache.spark.SparkException: Job aborted due to stage failure: Task 2 in stage 2.0 failed 1 times, most recent failure: Lost task 2.0 in stage 2.0 (TID 7, localhost, executor driver): java.lang.NullPointerException
	at java.util.Collections$UnmodifiableMap.<init>(Collections.java:1446)
	at java.util.Collections.unmodifiableMap(Collections.java:1433)
	at com.amazonaws.AmazonWebServiceRequest.getHandlerContext(AmazonWebServiceRequest.java:560)
	at com.amazonaws.DefaultRequest.<init>(DefaultRequest.java:104)
	at com.amazonaws.services.s3.AmazonS3Client.createRequest(AmazonS3Client.java:4112)
	at com.amazonaws.services.s3.AmazonS3Client.createRequest(AmazonS3Client.java:4097)
	at com.amazonaws.services.s3.AmazonS3Client.getObject(AmazonS3Client.java:1345)
	at geotrellis.spark.io.s3.AmazonS3Client.readRange(AmazonS3Client.scala:90)
	at geotrellis.spark.io.s3.util.S3RangeReader.readClippedRange(S3RangeReader.scala:48)
	at geotrellis.util.RangeReader$class.readRange(RangeReader.scala:36)
	at geotrellis.spark.io.s3.util.S3RangeReader.readRange(S3RangeReader.scala:38)
	at geotrellis.util.StreamingByteReader$$anonfun$1.apply(StreamingByteReader.scala:90)
	at geotrellis.util.StreamingByteReader$$anonfun$1.apply(StreamingByteReader.scala:90)
	at geotrellis.util.StreamingByteReader$Chunk.data(StreamingByteReader.scala:43)
	at geotrellis.util.StreamingByteReader$Chunk.buffer$lzycompute(StreamingByteReader.scala:48)
	at geotrellis.util.StreamingByteReader$Chunk.buffer(StreamingByteReader.scala:48)
	at geotrellis.util.StreamingByteReader.position(StreamingByteReader.scala:75)
	at geotrellis.raster.io.geotiff.reader.GeoTiffReader$.readGeoTiffInfo(GeoTiffReader.scala:369)
	at geotrellis.raster.io.geotiff.reader.GeoTiffReader$.readMultiband(GeoTiffReader.scala:175)
	at geotrellis.raster.io.geotiff.MultibandGeoTiff$.streaming(MultibandGeoTiff.scala:120)
	at geotrellis.spark.io.RasterReader$$anon$2.readWindow(RasterReader.scala:115)
	at geotrellis.spark.io.RasterReader$$anon$2.readWindow(RasterReader.scala:107)
	at geotrellis.spark.io.s3.S3GeoTiffRDD$$anonfun$apply$7.apply(S3GeoTiffRDD.scala:239)
	at geotrellis.spark.io.s3.S3GeoTiffRDD$$anonfun$apply$7.apply(S3GeoTiffRDD.scala:231)
	at scala.collection.Iterator$$anon$11.next(Iterator.scala:409)
	at scala.collection.Iterator$$anon$11.next(Iterator.scala:409)
	at scala.collection.Iterator$class.foreach(Iterator.scala:893)
	at scala.collection.AbstractIterator.foreach(Iterator.scala:1336)
	at scala.collection.TraversableOnce$class.reduceLeft(TraversableOnce.scala:185)
	at scala.collection.AbstractIterator.reduceLeft(Iterator.scala:1336)
	at org.apache.spark.rdd.RDD$$anonfun$reduce$1$$anonfun$15.apply(RDD.scala:1011)
	at org.apache.spark.rdd.RDD$$anonfun$reduce$1$$anonfun$15.apply(RDD.scala:1009)
	at org.apache.spark.SparkContext$$anonfun$33.apply(SparkContext.scala:1980)
	at org.apache.spark.SparkContext$$anonfun$33.apply(SparkContext.scala:1980)
	at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:87)
	at org.apache.spark.scheduler.Task.run(Task.scala:99)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:282)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:748)

Driver stacktrace:
	at org.apache.spark.scheduler.DAGScheduler.org$apache$spark$scheduler$DAGScheduler$$failJobAndIndependentStages(DAGScheduler.scala:1435)
	at org.apache.spark.scheduler.DAGScheduler$$anonfun$abortStage$1.apply(DAGScheduler.scala:1423)
	at org.apache.spark.scheduler.DAGScheduler$$anonfun$abortStage$1.apply(DAGScheduler.scala:1422)
	at scala.collection.mutable.ResizableArray$class.foreach(ResizableArray.scala:59)
	at scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:48)
	at org.apache.spark.scheduler.DAGScheduler.abortStage(DAGScheduler.scala:1422)
	at org.apache.spark.scheduler.DAGScheduler$$anonfun$handleTaskSetFailed$1.apply(DAGScheduler.scala:802)
	at org.apache.spark.scheduler.DAGScheduler$$anonfun$handleTaskSetFailed$1.apply(DAGScheduler.scala:802)
	at scala.Option.foreach(Option.scala:257)
	at org.apache.spark.scheduler.DAGScheduler.handleTaskSetFailed(DAGScheduler.scala:802)
	at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.doOnReceive(DAGScheduler.scala:1650)
	at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.onReceive(DAGScheduler.scala:1605)
	at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.onReceive(DAGScheduler.scala:1594)
	at org.apache.spark.util.EventLoop$$anon$1.run(EventLoop.scala:48)
	at org.apache.spark.scheduler.DAGScheduler.runJob(DAGScheduler.scala:628)
	at org.apache.spark.SparkContext.runJob(SparkContext.scala:1918)
	at org.apache.spark.SparkContext.runJob(SparkContext.scala:1981)
	at org.apache.spark.rdd.RDD$$anonfun$reduce$1.apply(RDD.scala:1025)
	at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:151)
	at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:112)
	at org.apache.spark.rdd.RDD.withScope(RDD.scala:362)
	at org.apache.spark.rdd.RDD.reduce(RDD.scala:1007)
	at geopyspark.geotrellis.RasterSummary$.collect(RasterSummary.scala:61)
	at geopyspark.geotrellis.ProjectedRasterLayer.tileToLayout(ProjectedRasterLayer.scala:67)
	at geopyspark.geotrellis.ProjectedRasterLayer.reproject(ProjectedRasterLayer.scala:82)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at py4j.reflection.MethodInvoker.invoke(MethodInvoker.java:244)
	at py4j.reflection.ReflectionEngine.invoke(ReflectionEngine.java:357)
	at py4j.Gateway.invoke(Gateway.java:280)
	at py4j.commands.AbstractCommand.invokeMethod(AbstractCommand.java:132)
	at py4j.commands.CallCommand.execute(CallCommand.java:79)
	at py4j.GatewayConnection.run(GatewayConnection.java:214)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.NullPointerException
	at java.util.Collections$UnmodifiableMap.<init>(Collections.java:1446)
	at java.util.Collections.unmodifiableMap(Collections.java:1433)
	at com.amazonaws.AmazonWebServiceRequest.getHandlerContext(AmazonWebServiceRequest.java:560)
	at com.amazonaws.DefaultRequest.<init>(DefaultRequest.java:104)
	at com.amazonaws.services.s3.AmazonS3Client.createRequest(AmazonS3Client.java:4112)
	at com.amazonaws.services.s3.AmazonS3Client.createRequest(AmazonS3Client.java:4097)
	at com.amazonaws.services.s3.AmazonS3Client.getObject(AmazonS3Client.java:1345)
	at geotrellis.spark.io.s3.AmazonS3Client.readRange(AmazonS3Client.scala:90)
	at geotrellis.spark.io.s3.util.S3RangeReader.readClippedRange(S3RangeReader.scala:48)
	at geotrellis.util.RangeReader$class.readRange(RangeReader.scala:36)
	at geotrellis.spark.io.s3.util.S3RangeReader.readRange(S3RangeReader.scala:38)
	at geotrellis.util.StreamingByteReader$$anonfun$1.apply(StreamingByteReader.scala:90)
	at geotrellis.util.StreamingByteReader$$anonfun$1.apply(StreamingByteReader.scala:90)
	at geotrellis.util.StreamingByteReader$Chunk.data(StreamingByteReader.scala:43)
	at geotrellis.util.StreamingByteReader$Chunk.buffer$lzycompute(StreamingByteReader.scala:48)
	at geotrellis.util.StreamingByteReader$Chunk.buffer(StreamingByteReader.scala:48)
	at geotrellis.util.StreamingByteReader.position(StreamingByteReader.scala:75)
	at geotrellis.raster.io.geotiff.reader.GeoTiffReader$.readGeoTiffInfo(GeoTiffReader.scala:369)
	at geotrellis.raster.io.geotiff.reader.GeoTiffReader$.readMultiband(GeoTiffReader.scala:175)
	at geotrellis.raster.io.geotiff.MultibandGeoTiff$.streaming(MultibandGeoTiff.scala:120)
	at geotrellis.spark.io.RasterReader$$anon$2.readWindow(RasterReader.scala:115)
	at geotrellis.spark.io.RasterReader$$anon$2.readWindow(RasterReader.scala:107)
	at geotrellis.spark.io.s3.S3GeoTiffRDD$$anonfun$apply$7.apply(S3GeoTiffRDD.scala:239)
	at geotrellis.spark.io.s3.S3GeoTiffRDD$$anonfun$apply$7.apply(S3GeoTiffRDD.scala:231)
	at scala.collection.Iterator$$anon$11.next(Iterator.scala:409)
	at scala.collection.Iterator$$anon$11.next(Iterator.scala:409)
	at scala.collection.Iterator$class.foreach(Iterator.scala:893)
	at scala.collection.AbstractIterator.foreach(Iterator.scala:1336)
	at scala.collection.TraversableOnce$class.reduceLeft(TraversableOnce.scala:185)
	at scala.collection.AbstractIterator.reduceLeft(Iterator.scala:1336)
	at org.apache.spark.rdd.RDD$$anonfun$reduce$1$$anonfun$15.apply(RDD.scala:1011)
	at org.apache.spark.rdd.RDD$$anonfun$reduce$1$$anonfun$15.apply(RDD.scala:1009)
	at org.apache.spark.SparkContext$$anonfun$33.apply(SparkContext.scala:1980)
	at org.apache.spark.SparkContext$$anonfun$33.apply(SparkContext.scala:1980)
	at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:87)
	at org.apache.spark.scheduler.Task.run(Task.scala:99)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:282)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	... 1 more
```

--- a/geopyspark-backend/geotrellis/build.sbt
+++ b/geopyspark-backend/geotrellis/build.sbt
@@ -20,6 +20,8 @@ libraryDependencies ++= Seq(
   "net.sf.py4j"           % "py4j"                               % "0.10.5"
 )
 
+dependencyOverrides += "com.amazonaws" % "aws-java-sdk-s3" % "1.11.105"
+
 assemblyMergeStrategy in assembly := {
   case "reference.conf" => MergeStrategy.concat
   case "application.conf" => MergeStrategy.concat

--- a/geopyspark-backend/geotrellis/src/main/scala/geopyspark/geotrellis/io/geotiff/GeoTiffRDD.scala
+++ b/geopyspark-backend/geotrellis/src/main/scala/geopyspark/geotrellis/io/geotiff/GeoTiffRDD.scala
@@ -58,6 +58,14 @@ object GeoTiffRDD {
         else
           None
 
+      val partitionBytes =
+        if (javaMap.contains("partition_bytes") && !intMap.contains("max_tile_size"))
+          Some(javaMap.get("partition_bytes").asInstanceOf[Long])
+        else if (!intMap.contains("max_tile_size"))
+          default.partitionBytes
+        else
+          None
+
       val getS3Client: () => S3Client =
         stringMap.get("s3_client") match {
           case Some(client) =>
@@ -76,6 +84,7 @@ object GeoTiffRDD {
         timeFormat = stringMap.getOrElse("time_format", default.timeFormat),
         maxTileSize = intMap.get("max_tile_size"),
         numPartitions = intMap.get("num_partitions"),
+        partitionBytes = partitionBytes,
         chunkSize = intMap.get("chunk_size"),
         getS3Client = getS3Client)
     }

--- a/geopyspark/geotrellis/geotiff.py
+++ b/geopyspark/geotrellis/geotiff.py
@@ -40,6 +40,18 @@ def get(layer_type,
         num_partitions (int, optional): The number of repartitions Spark
             will make when the data is repartitioned. If ``None``, then the
             data will not be repartitioned.
+        partition_bytes (int, optional): The desired number of bytes per partition.
+            This is will ensure that at least one item is assigned for each partition.
+            If ``None`` and ``max_tile_size`` is not set, then the default size per
+            partition is 128 Mb.
+
+            Note:
+                This option is only available when reading from S3.
+
+            Note:
+                This option is incompatible with the ``max_tile_size`` option.
+                If both are set, then ``max_tile_size`` will be used instead of
+                ``partition_bytes``.
         chunk_size (int, optional): How many bytes of the file should be
             read in at a time. If ``None``, then files will be read in 65536
             byte chunks.

